docs/design/coreclr/botr/memory-model.md

```diff
@@ -0,0 +1,248 @@
+
+# CLR memory model
+
+## ECMA vs. CLR memory models.
+ECMA 335 standard defines a very weak memory model. After two decades the desire to have a flexible model did not result in considerable benefits due to hardware being more strict. On the other hand programming against ECMA model requires extra complexity to handle scenarios that are hard to comprehend and not possible to test.
+
+In the course of multiple releases CLR implementation settled around a memory model that is a practical compromise between what can be implemented efficiently on the current hardware, while staying reasonably approachable by the developers. This document rationalizes the invariants provided and expected by the CLR runtime in its current implementation with expectation of that being carried to future releases.
+
+## Alignment
+When managed by CLR runtime, variables of built-in primitive types are *properly aligned* according to the data type size. This applies to both heap and stack allocated memory.
+
+1-byte, 2-byte, 4-byte variables are stored at 1-byte, 2-byte, 4-byte boundary, respectively.  
+8-byte variables are 8-byte aligned on 64 bit platforms.  
+Native-sized integer types and pointers have alignment that matches their size on the given platform.  
+
+## Atomic memory accesses.
+Memory accesses to *properly aligned* data of primitive types are always atomic. The value that is observed is always a result of complete read and write operations.
```

Comment:
Do we also want to mention that reading and storing object references and unmanaged pointers is atomic?

Reply:
Object references are always aligned. We can mention that (if it is already not mentioned).
But unmanaged pointers are just numbers with derefeence operation defined. They may not be aligned, then reading/storing is not necessarily atomic.

Reply:
Ah, you mean the pointers themselves. I guess, they are "properly aligned" and thus read/writes are atomic.

Reply:
I meant the value of unmanaged pointer. Unmanaged pointer is not a primitive type, so the current wording does not cover it.

Reply:
Yes, it may be worth mentioning this. It follows from the "properly aligned" but it is a by far most common case.